src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -292,7 +292,14 @@ private function getAutowiredReference(TypedReference $reference)
         }
 
         if (isset($this->types[$type])) {
-            @trigger_error(sprintf('Autowiring services based on the types they implement is deprecated since Symfony 3.3 and won\'t be supported in version 4.0. You should %s the "%s" service to "%s" instead.', isset($this->types[$this->types[$type]]) ? 'alias' : 'rename (or alias)', $this->types[$type], $type), E_USER_DEPRECATED);
+            $message = 'Autowiring services based on the types they implement is deprecated since Symfony 3.3 and won\'t be supported in version 4.0.';
+            if ($aliasSuggestion = $this->getAliasesSuggestionForType($type = $reference->getType())) {
+                $message .= ' '.$aliasSuggestion;
+            } else {
+                $message .= sprintf(' You should %s the "%s" service to "%s" instead.', isset($this->types[$this->types[$type]]) ? 'alias' : 'rename (or alias)', $this->types[$type], $type);
+            }
+
+            @trigger_error($message, E_USER_DEPRECATED);
```

Comment:
I'm mixed on this one. The current message tells about the *target* service. The new message about the *source* one, but gives little info about it.
Even if we fix that, it means we won't be able to aggregate deprecation notices as we usually do. People will get a bunch of different message for different sources, all related to the same target service in fact.